rclpy/src/rclpy/qos.cpp

```diff
@@ -0,0 +1,135 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Include pybind11 before rclpy_common/handle.h includes Python.h
+#include <pybind11/pybind11.h>
+
+#include <rcl/rcl.h>
+#include <rcl/time.h>
+#include <rmw/types.h>
+
+#include <cstring>
+#include <memory>
+#include <string>
+
+#include "rclpy_common/common.h"
+#include "rclpy_common/handle.h"
+
+#include "qos.hpp"
+
+namespace rclpy
+{
+
+namespace
+{
+
+// Fill a given rmw_time_t from an rclpy.duration.Duration instance.
+void
+_convert_py_duration_to_rmw_time(py::capsule pyduration, rmw_time_t * out_time)
+{
+  assert(out_time != nullptr);
+  auto duration = static_cast<rcl_duration_t *>(pyduration);
+  out_time->sec = RCL_NS_TO_S(duration->nanoseconds);
+  out_time->nsec = duration->nanoseconds % (1000LL * 1000LL * 1000LL);
+}
+
+}  // namespace
+
+py::capsule
+convert_from_py_qos_policy(
+  int qos_history,
+  int qos_depth,
+  int qos_reliability,
+  int qos_durability,
+  py::capsule pyqos_lifespan,
+  py::capsule pyqos_deadline,
+  int qos_liveliness,
+  py::capsule pyqos_liveliness_lease_duration,
+  bool avoid_ros_namespace_conventions)
+{
+  auto qos_profile =
+    std::unique_ptr<rmw_qos_profile_t, decltype(&PyMem_Free)>(
+    static_cast<rmw_qos_profile_t *>(
+      PyMem_Malloc(sizeof(rmw_qos_profile_t))), PyMem_Free);
+  if (!qos_profile) {
+    throw std::bad_alloc();
+  }
+  // Set to default so that we don't use uninitialized data if new fields are added
+  *qos_profile = rmw_qos_profile_default;
+
+  // Overwrite defaults with passed values
+  qos_profile->history = static_cast<rmw_qos_history_policy_t>(qos_history);
+  qos_profile->depth = qos_depth;
+  qos_profile->reliability = static_cast<rmw_qos_reliability_policy_t>(qos_reliability);
+  qos_profile->durability = static_cast<rmw_qos_durability_policy_t>(qos_durability);
+
+  _convert_py_duration_to_rmw_time(pyqos_lifespan, &(qos_profile->lifespan));
+  _convert_py_duration_to_rmw_time(pyqos_deadline, &(qos_profile->deadline));
+
+  qos_profile->liveliness = static_cast<rmw_qos_liveliness_policy_t>(qos_liveliness);
+
+  _convert_py_duration_to_rmw_time(
+    pyqos_liveliness_lease_duration,
+    &(qos_profile->liveliness_lease_duration));
+
+  qos_profile->avoid_ros_namespace_conventions = avoid_ros_namespace_conventions;
+
+  return py::capsule(
+    qos_profile.release(), "rmw_qos_profile_t",
+    [](PyObject * pycapsule_c) {
+      PyMem_Free(PyCapsule_GetPointer(pycapsule_c, "rmw_qos_profile_t"));
+    });
+}
+
+py::dict
+convert_to_py_qos_policy(py::capsule pyqos_profile)
+{
+  auto qos_profile = static_cast<rmw_qos_profile_t *>(pyqos_profile);
```

Comment:
Also recommend checking the capsule name here

Reply:
See https://github.com/ros2/rclpy/pull/736/commits/e0d364465df44d78bd4fade06e65bb1be177467f.